File: src_tf/demo_cpp_tf/src/tf_listener.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp> //提供消息接口
#include <tf2/LinearMath/Quaternion.h> //四元数类
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp> //四元数与消息的转换
#include <tf2_ros/transform_listener.h> //坐标容器
#include <tf2_ros/buffer.h>
#include <tf2/utils.h>
#include <chrono>

using namespace std::chrono_literals;


class TFListenerNode : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
public:
    TFListenerNode() : Node("tf_listener")
    {
        this->tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        this->tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, this);
        this->timer_ = this->create_wall_timer(
            1s,
            std::bind(&TFListenerNode::get_transform, this)
        );
    }

    void get_transform()
    {
        geometry_msgs::msg::TransformStamped transform;
        try
        {
            transform = this->tf_buffer_->lookupTransform("base_link", "target_point", this->get_clock()->now(),
            rclcpp::Duration::from_seconds(1.0));
            RCLCPP_INFO(this->get_logger(), "Translation: [%.2f, %.2f, %.2f]", 
                transform.transform.translation.x,
                transform.transform.translation.y,
                transform.transform.translation.z);

            double roll, pitch, yaw;
            tf2::getEulerYPR(transform.transform.rotation, yaw, pitch, roll);
            RCLCPP_INFO(this->get_logger(), "Rotation: [Roll: %.2f, Pitch: %.2f, Yaw: %.2f]", 
                roll, pitch, yaw);
            
        }
        catch (const std::exception &ex)
        {
            RCLCPP_WARN(this->get_logger(), "%s", ex.what());
        }
    }


};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFListenerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src_tf/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer #坐标监听器
from tf_transformations import euler_from_quaternion #欧拉角转四元数
import math 

class TFBroadCaster(Node):
    def __init__(self):
        super().__init__('tf_broadcaster')
        self.buffer_ = Buffer()
        self.listener_ = TransformListener(self.buffer_, self)
        self.timer_ = self.create_timer(1.0, self.get_transform) 

    #查询坐标关系
    def get_transform(self):
        try:
            result = self.buffer_.lookup_transform('base_link', 'bottle_link',
                     rclpy.time.Time(seconds=0.0), rclpy.time.Duration(seconds=1.0))
            transform = result.transform
            self.get_logger().info(f'平移: {transform.translation}')
            self.get_logger().info(f'旋转: {transform.rotation}')
            self.get_logger().info(f'欧拉角: {euler_from_quaternion([transform.rotation.x, transform.rotation.y, transform.rotation.z, transform.rotation.w])}')
        except Exception as e:
            self.get_logger().info(f'Exception {e}')
            return

def main():
    rclpy.init()
    node = TFBroadCaster()
    rclpy.spin(node)
    rclpy.shutdown()
